### PR TITLE
verify response body with expected keywords

### PR DESCRIPTION
**Pull request by zubairk14**
Sometimes, HTTP calls to endpoints with _invalid_ credentials return a `200` status, but without the actual access token (or actual access to the resource). Verifying response body with expected keywords helps mitigate that source of false positives.

### CODE REVIEW

File: pkg/common/utils.go [hunk 1]
```diff
@@ -17,3 +23,44 @@ func RemoveStringSliceItem(item string, slice *[]string) {
...
+		if err != nil {
+			return false, err
+		}
+		return found, nil
+	}
+
+	return false, nil
+}
+
+func containsSubstring(reader io.ReadCloser, target string) (bool, error) {
```
Comment (ahrav) [inline pkg/common/utils.go:45]: suggestion: You can probably even use the `bufio.NewScanner` here. It's a little easier to read imo.
```
func containsSubstring(reader io.Reader, substring string) (bool, error) {
	scanner := bufio.NewScanner(reader)
	for scanner.Scan() {
		if strings.Contains(scanner.Text(), substring) {
			return true, nil
		}
	}
	if err := scanner.Err(); err != nil {
		return false, err
	}
	return false, nil
}
```

File: pkg/common/utils_test.go [hunk 2]
```diff
@@ -1,7 +1,9 @@
 package common
 
 import (
+	"io/ioutil"
```
Comment (ahrav) [inline pkg/common/utils_test.go:4]: suggestion: io/ioutil is deprecated, so we should just use io

File: pkg/common/utils_test.go [hunk 3]
```diff
@@ -74,3 +76,31 @@ func TestRemoveItem(t *testing.T) {
 		}
 	}
 }
+
+// Test ParseResponseForKeywords with a reader that contains the keyword and a reader that doesn't.
+func TestParseResponseForKeywords(t *testing.T) {
```
Comment (ahrav) [inline pkg/common/utils_test.go:81]: suggestion: Since we have multiple test cases with the exact same setup we might benefit from a table-driven test. We could also maybe throw in some edge cases?

https://google.github.io/styleguide/go/best-practices#tests
```
If inputs are similar, consider unifying them into a table-driven test while keeping the logic inlined in the loop. This helps to avoid repetition while keeping the validation and failure in the Test.
```

Ex:
```
func TestParseResponseForKeywords(t *testing.T) {
	testCases := []struct {
		name     string
		input    string
		keywords []string
		expected bool
	}{
		{
			name:     "Should find keyword",
			input:    "ey: abc",
			keywords: []string{"ey"},
			expected: true,
		},
		{
			name:     "Should not find keyword",
			input:    "fake response",
			keywords: []string{"ey"},
			expected: false,
		},
		{
			name:     "Empty string",
			input:    "",
			keywords: []string{"ey"},
			expected: false,
		},
		{
			name:     "Keyword at end",
			input:    "abc ey",
			keywords: []string{"ey"},
			expected: true,
		},
		{
			name:     "Keyword at start",
			input:    "ey abc",
			keywords: []string{"ey"},
			expected: true,
		},
	}

	for _, tc := range testCases {
		t.Run(tc.name, func(t *testing.T) {
			testReader := strings.NewReader(tc.input)
			testReadCloser := io.NopCloser(testReader)
			found, err := ParseResponseForKeywords(testReadCloser, tc.keywords)

			if err != nil {
				t.Errorf("Error: %v", err)
			}

			if found != tc.expected {
				t.Errorf("Expected %v, got %v", tc.expected, found)
			}
		})
	}
}
```

File: pkg/detectors/docusign/docusign.go [hunk 4]
```diff
@@ -83,25 +81,14 @@ func (s Scanner) FromData(ctx context.Context, verify bool, data []byte) (result
 					return nil, errors.WrapPrefix(err, "Error making request", 0)
 				}
 
-				// Read the response body
-				body, err := io.ReadAll(res.Body)
+				verifiedBodyResponse, err := common.ParseResponseForKeywords(res.Body, []string{"ey"})
```
Comment (ahrav) [inline pkg/detectors/docusign/docusign.go:84]: suggestion: Linter noticed the response body is not closed, looks like it got removed in the refactor. We should throw that back in here to prevent any resource leaks.

**Review by ahrav — APPROVED** [review]

File: pkg/common/utils.go (hunk 1 above)
Comment (zubairk14) [reply pkg/common/utils.go:45]: thanks for the suggestion!

File: pkg/common/utils_test.go (hunk 3 above)
Comment (zubairk14) [reply pkg/common/utils_test.go:81]: ah true

File: pkg/common/utils.go [hunk 5]
```diff
@@ -17,3 +23,44 @@ func RemoveStringSliceItem(item string, slice *[]string) {
 		}
 	}
 }
+
+// ParseResponseForKeywords parses the response from detector verification calls for expected keywords in the response.
+func ParseResponseForKeywords(reader io.ReadCloser, keywords []string) (bool, error) {
```
Comment (dustin-decker) [inline pkg/common/utils.go:28]: This won't perform as expected because reading from the body consumes the reader. That means that once you start reading it, you won't be able to read those parts again. Since we only need to check the response for one substring anyway, I recommend removing this function and just renaming `containsSubstring` to `ResponseContainsSubstring`

**Review by dustin-decker — CHANGES_REQUESTED** [review]

**Review by dustin-decker — APPROVED** [review]
